pkg/arvo/lib/eth-abi.hoon

```diff
@@ -0,0 +1,168 @@
+      outputs=(list @tas)
+  ==
++$  event
+  $:  input-sol=(list event-input)
+      input-hoon=(list @tas)
+  ==
++$  event-input  [type=@t indexed=?]
++$  contract-raw
+  $:  name=@t
+      entries=(list entry-raw)
+  ==
++$  entry-raw
+  $%  [%function p=function-raw]
+      [%event p=event-raw]
+  ==
++$  function-raw
+  $:  name=@t
+      inputs=(list @tas)
+      outputs=(list @tas)
+      mut=?
+      pay=?
+  ==
++$  event-raw
+  $:  name=@t
+      inputs=(list event-input)
+  ==
+::
+++  parse-contract
+  |=  jon=json
+  =/  =contract-raw
+    =,  dejs:format
+    %.  jon  %-  ot
+    :~  ['contractName' so]
+        [%abi parse-abi]
+    ==
+  ^-  contract
+  :^    name.contract-raw
+      (get-write-functions entries.contract-raw)
+    (get-read-functions entries.contract-raw)
+  (get-events entries.contract-raw)
+::
+++  parse-abi
+  |=  jon=json
+  =,  dejs:format
+  ^-  (list entry-raw)
+  %.  jon  %-  ar
+  |=  jan=json
+  ?>  ?=([$o *] jan)
+  =/  typ  (so (~(got by p.jan) 'type'))
+  %.  jan
+  ?+    typ  !!
+      %function
+    |=  jun=json
+    :-  %function
+    ^-  function-raw
+    %.  jun
+    %-  ot
+    =/  extract-func-field
+      |=  jyn=json
+      ?>  ?=([$o *] jyn)
+      ^-  @tas
+      (so (~(got by p.jyn) 'type'))
+    :~  [%name so]
+        [%inputs (ar extract-func-field)]
+        [%outputs (ar extract-func-field)]
+        [%constant |=(jen=json !(bo jen))]
+        [%payable bo]
+    ==
+  ::
+      %event
+    |=  jun=json
+    :-  %event
+    ^-  event-raw
+    %.  jun
+    %-  ot
+    :~  [%name so]
+    ::
+      :-  %inputs
+      %-  ar
+      %-  ot
+      :~  [%type so]
+          [%indexed bo]
+      ==
+    ==
+  ==
+::
+++  get-raw-functions
+  |=  abi=(list entry-raw)
+  ^-  (list function-raw)
+  %+  roll  abi
+  |=  [e=entry-raw fs=(list function-raw)]
+    ?.  ?=([$function *] e)  fs
+    [p.e fs]
+::
+++  get-raw-events
+  |=  abi=(list entry-raw)
+  ^-  (list event-raw)
+  %+  roll  abi
+  |=  [e=entry-raw fs=(list event-raw)]
+    ?.  ?=([$event *] e)  fs
+    [p.e fs]
+::
+++  get-events
+  |=  abi=(list entry-raw)
+  ^-  (map @ux event)
+  %+  roll  (get-raw-events abi)
+  |=  [e=event-raw es=(map @ux event)]
+  =/  typs  (turn inputs.e |=(e=event-input type.e))
+  %+  ~(put by es)
+    (get-hash name.e typs)
+    [inputs.e (parse-types typs)]
+::
+++  get-write-functions
+  |=  abi=(list entry-raw)
+  ^-  (map @tas function)
+  %+  roll  (get-raw-functions abi)
+  |=  [f=function-raw fs=(map @tas function)]
+  ?.  mut.f  fs
+  %+  ~(put by fs)  name.f
+  :-  inputs.f  :-  (parse-types inputs.f)  outputs.f
+::
+++  get-read-functions
+  |=  abi=(list entry-raw)
+  ^-  (map @tas function)
+  %+  roll  (get-raw-functions abi)
+  |=  [f=function-raw fs=(map @tas function)]
+  ?:  mut.f  fs
+  %+  ~(put by fs)  name.f
+  :-  inputs.f  :-  (parse-types inputs.f)  outputs.f
+::
+++  parse-types
+  |=  typs=(list @t)
+  %+  turn  typs
+  |=  typ=@tas
+  ?+  (crip (scag 3 (trip typ)))
```

Comment:
Kind of surprised to see an expression in the `?+` condition, but also I don't understand why we `(scag 3` the type here? Is this to avoid dealing with different `int` and co sizes, like `int32`, `uint64` etc? Whatever the reason, it should probably be annotated.